drivers/sensor/dht20/dht20.c

```diff
@@ -0,0 +1,359 @@
+	} else if (chan == SENSOR_CHAN_HUMIDITY) {
+		dht20_rh_convert(val, data->rh_sample);
+	} else {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int dht20_init(const struct device *dev)
+{
+	const struct dht20_config *cfg = dev->config;
+
+	if (!i2c_is_ready_dt(&cfg->bus)) {
+		LOG_ERR("I2C dev %s not ready", cfg->bus.bus->name);
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static const struct sensor_driver_api dht20_driver_api = {.sample_fetch = dht20_sample_fetch,
+							  .channel_get = dht20_channel_get};
+
+#define DT_DRV_COMPAT aosong_dht20
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+
+#define DEFINE_DHT20(n)                                                                            \
+	static struct dht20_data dht20_data_##n;                                                   \
+                                                                                                   \
+	static const struct dht20_config dht20_config_##n = {.bus = I2C_DT_SPEC_INST_GET(n)};      \
+                                                                                                   \
+	SENSOR_DEVICE_DT_INST_DEFINE(n, dht20_init, NULL, &dht20_data_##n, &dht20_config_##n,      \
+				     POST_KERNEL, CONFIG_SENSOR_INIT_PRIORITY, &dht20_driver_api);
+
+DT_INST_FOREACH_STATUS_OKAY(DEFINE_DHT20)
+
+#endif
+#undef DT_DRV_COMPAT
+
+#define DT_DRV_COMPAT aosong_aht20
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+
+#define DEFINE_AHT20(n)                                                                            \
+	static struct dht20_data aht20_data_##n;                                                   \
+                                                                                                   \
+	static const struct dht20_config aht20_config_##n = {.bus = I2C_DT_SPEC_INST_GET(n)};      \
+                                                                                                   \
+	SENSOR_DEVICE_DT_INST_DEFINE(n, dht20_init, NULL, &aht20_data_##n, &aht20_config_##n,      \
+				     POST_KERNEL, CONFIG_SENSOR_INIT_PRIORITY, &dht20_driver_api);
+
+DT_INST_FOREACH_STATUS_OKAY(DEFINE_AHT20)
+
+#endif
+#undef DT_DRV_COMPAT
+
+#define DT_DRV_COMPAT aosong_am2301b
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+
+#define DEFINE_AM2301B(n)                                                                          \
+	static struct dht20_data am2301b_data_##n;                                                 \
+                                                                                                   \
+	static const struct dht20_config am2301b_config_##n = {.bus = I2C_DT_SPEC_INST_GET(n)};    \
+                                                                                                   \
+	SENSOR_DEVICE_DT_INST_DEFINE(n, dht20_init, NULL, &am2301b_data_##n, &am2301b_config_##n,  \
+				     POST_KERNEL, CONFIG_SENSOR_INIT_PRIORITY, &dht20_driver_api);
+
+DT_INST_FOREACH_STATUS_OKAY(DEFINE_AM2301B)
```

Comment:
I am not a DT macro expert but is there no way to make this just one block? (if DT_HAS_COMPAT_STATUS_OKAY(variant1) or DT_HAS_COMPAT_STATUS_OKAY(variant2) or DT_HAS_COMPAT_STATUS_OKAY(variant3)) and therefore get rid of the duplication?

Reply:
Makes sense. Summoning @fabiobaltieri just in case he has a brilliant idea, if not we'll just leave it at that I guess :)
Thanks!

Reply:
No magic, this is just how things are at the moment, any idea on how to improve the situation for drivers supporting multiple compatibles would be very welcome but at this point this is how it's done.

Reply:
Thank you, @fabiobaltieri